src/v/kafka/server/handlers/handler_probe.cc

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "kafka/server/handlers/handler_probe.h"
+
+#include "config/configuration.h"
+#include "kafka/server/handlers/handler_interface.h"
+#include "prometheus/prometheus_sanitize.h"
+
+#include <seastar/core/metrics.hh>
+
+#include <memory>
+
+namespace kafka {
+
+handler_probe_manager::handler_probe_manager()
+  : _probes((size_t)max_handler_api_key + 2)
+  , _unknown_handler_index((size_t)max_handler_api_key + 1) {
+    for (size_t i = 0; i < _probes.size(); i++) {
+        auto key = api_key{i};
+
+        if (handler_for_key(key) || i == _unknown_handler_index) {
+            _probes[i] = std::make_unique<handler_probe>(key);
+            _probes.at(i)->setup_metrics();
```

Comment:
this looks a lot like the existing handler interface. can we attach a probe to the existing `struct handler_info` in `handler_interface`? or other handler related types?

Reply:
I don't think `handler_interface` would be a good place for those: the underlying objects are const or constexpr singletons with unlimited lifetime and shared across shards. They are basically an immutable collection of facts & methods associated with each request type, the runtime polymorphic equivalent of the templated handler types. Putting mutable state like the probe in there would be a big change and seems to be the wrong cardinality and lifetime (e.g., it would mix metrics from unrelated applications running in the same process).

I see what you are saying though: we want something which is per-kafka-server-per-handler and the probe would naturally live there. Let's check if we have something like that already. I think the existing approach is usually that handlers just stash they stuff as members on the server object then access it with something like `request_context->connection()->server()`: maybe we are missing a `handler_context` object which is shared persistent state across all requests of the same type? That's where fetch would put their fetch session cache, etc.

Reply:
>Putting mutable state like the probe in there would be a big change and seems to be the wrong cardinality and lifetime

ah yes, of course you're right about that, it would be an odd match.

>maybe we are missing a handler_context object which is shared persistent state across all requests of the same type?

Yeh, but I think the current approach with the ` kafka::server::handler_probe_manager _handler_probes;` works well too. The larger picture is that we are slowly refactoring the kafka::server hence the thought about dangling off the existing per-handler abstraction. But the manager approach is not going to cause any issues either.